feat(scim): Provisioning org member roles

AniketDas-Tekky opened the pull request:
This PR modifies the SCIM org member provisioning by allowing for the caller to provide a `sentryOrgRole`. The API will attempt to set the role for the new user. This change also includes some auxillary changes for abstraction and test coverage.
- A new SCIMApiError drf exception automatically adds the schema to responses
- SCIM tests now log in as the scim token instead of the owner.

Commit:
feat(scim): Provisioning org member roles

This PR modifies the SCIM org member provisioning by allowing for the caller to provide a `sentryOrgRole`. The API will attempt to set the role for the new user.
This change also includes some auxillary changes for abstraction and test coverage.
- A new SCIMApiError drf exception automatically adds the schema to responses
- SCIM tests now log in as the scim token instead of the owner.

AniketDas-Tekky commented on src/sentry/scim/endpoints/members.py:
```diff
@@ -221,7 +224,7 @@ def patch(self, request: Request, organization, member):
                 return Response(status=204)
             else:
-                return Response(SCIM_400_INVALID_PATCH, status=400)
+                raise SCIMApiError(detail=SCIM_400_INVALID_PATCH)
```
Technically this is a changed response since the old one didn't include the schema

AniketDas-Tekky commented on src/sentry/scim/endpoints/members.py:
```diff
@@ -370,16 +373,29 @@ def post(self, request: Request, organization) -> Response:
+        else:
+            role = organization.default_role
+
+        allowed_roles = roles.get_all()
```
This should be safe since the POST endpoint requires "member:admin" permissions.

RyanSkonnord replied:
Can you elaborate on this? Are you saying that, since the endpoint itself requires the `"member:admin"` scope, anyone who is accessing it must have the highest clearance to set all roles?

If so, the Manager role also has `"member:admin"`. Just from eyeballing this, it looks as though this would would allow a Manager to promote to Owner, which is a bug.

Secondly, even if that's not a bug, I think it's good to avoid hard-coding assumptions that certain scopes do or don't belong to certain roles. Role definitions are config values that can change over time, plus `UserPermission`s could mess it up unexpectedly.

AniketDas-Tekky replied:
So the way we prevent the self-promotion today is by doing a priority comparison between the org member's role and all the roles on the server; only allowing role assignment on roles with lower or equal priority.

The problem here is that the "user" assigning a member's role is an IDP integration and it doesn't have an org membership. So if we need to limit the roles that the integration can provision, we need to figure out a way to assign a role to the integration.

AniketDas-Tekky replied:
Also, need to find a place to assign the role to the integration

RyanSkonnord replied:
I see. Am I right in thinking that you only need to be a Manager to set up the integration? Is it a problem if a rogue Manager could set up a SCIM integration for the purpose of privilege-escalating a new account as an Owner? Or are we generally assuming that the pipeline of incoming SCIM data is trusted/controlled enough to give it unlimited discretion over all org memberships? (These might be product questions.)

JoshFerge replied:
yeah, im curious how the manager -> owner privilege escalation can go now, as i see you can set Owner as the default role. wonder if we should make it so owner can't be a default role and close this gap.

AniketDas-Tekky replied:
So we're gonna limit it to `member` and `billing` on SaaS. For here though, i'm gonna default to allowing every role.

JoshFerge replied:
i could see admin being useful as well / not super dangeorus but I'll leave it up to the team to figure out!

AniketDas-Tekky commented on src/sentry/scim/endpoints/members.py:
```diff
@@ -390,8 +406,8 @@ def post(self, request: Request, organization) -> Response:
                 # our error and if so, return a 409 so the scim IDP knows how to handle
-                raise ConflictError(detail=SCIM_409_USER_EXISTS)
-            return Response(serializer.errors, status=400)
+                raise SCIMApiError(detail=SCIM_409_USER_EXISTS, status_code=409)
```
feeling mixed about this, we're trading off automatically setting the status code for automatically setting the schema.

JoshFerge replied:
this seems cleaner to me!

AniketDas-Tekky commented on src/sentry/testutils/cases.py:
```diff
@@ -1914,7 +1915,8 @@ def setUp(self, provider="dummy"):
         self.auth_provider.enable_scim(self.user)
         self.auth_provider.save()
-        self.login_as(user=self.user)
+        self.scim_user = ApiToken.objects.get(token=self.auth_provider.get_scim_token()).user
```
This feels ugly. Is there a better way to get the auth_provider user?

RyanSkonnord replied:
Not really. Ideally we could do something like
```python
(self.scim_user,) = ApiToken.objects.filter(token=self.auth_provider.get_scim_token()).values_list("user", flat=True)
```
but that just returns the foreign key, not the `User` object. There are a few other ways to slice it but, considering that we don't need to optimize performance for `testutils`, I think what you've written is the cleanest option.

JoshFerge replied:
you could make a helper method that uses the token directly e.g. https://github.com/getsentry/sentry/blob/509241d00e459b2509c12d2d4a92bbca03a4873e/tests/sentry/api/endpoints/test_project_grouping_configs.py#L19

RyanSkonnord commented on tests/sentry/api/endpoints/test_scim_user_index.py:
```diff
@@ -69,6 +71,62 @@ def test_post_users_already_exists(self):
+                "schemas": ["urn:ietf:params:scim:api:messages:2.0:Error"],
+                "detail": "Invalid organization role.",
+            }
+
```
I'd like to see at least one test case that returns a `SCIM_400_INVALID_ORGROLE` status, where a user (who otherwise has the right permission to access the endpoint) is stopped from assigning a role above their own.

If I'm wrong about the Manager/Owner thing above and this is impossible under the current role structure, you could force the corner case with `override_settings` on the default role definitions. Or, if we can commit to `SCIM_400_INVALID_ORGROLE` being a logical impossibility, I'd suggest deleting that status and just raising a hard exception (with 500 response) instead.

JoshFerge commented on src/sentry/scim/endpoints/members.py:
```diff
@@ -370,16 +373,29 @@ def post(self, request: Request, organization) -> Response:
+            features.has("organizations:scim-orgmember-roles", organization, actor=None)
+            and "sentryOrgRole" in request.data
+        ):
+            role = request.data.get("sentryOrgRole").lower()
```
do you have to do `.get` if you've already validated `sentryOrgRole` is in `data`?

AniketDas-Tekky replied:
nope! force of habit.

JoshFerge commented on src/sentry/scim/endpoints/members.py:
```diff
@@ -370,16 +373,29 @@ def post(self, request: Request, organization) -> Response:
         - The API also does not support setting secondary emails.
+        - you can provide a sentryOrgRole to provision alongside the new user
         """
 
```
should we add this key to the example above? any other docs that need to be updated?

AniketDas-Tekky replied:
I actually dropped this change from the docstring. I don't want this to show up in public docs before it's ready.

JoshFerge replied:
👍🏼

JoshFerge reviewed (commented):
lgtm besides Qs about "owner" role privilege etc!

Commit:
Applied review comments and also new edge case for allowed roles

We will only allow member or billing users to be provisioned on SaaS, but here we'll allow any by default.
This is user configurable to be a subset

JoshFerge commented on src/sentry/scim/endpoints/members.py:
```diff
@@ -373,21 +373,23 @@ def post(self, request: Request, organization) -> Response:
         if (
             features.has("organizations:scim-orgmember-roles", organization, actor=None)
             and "sentryOrgRole" in request.data
+            and request.data["sentryOrgRole"]
```
what is this line checking?

AniketDas-Tekky replied:
if it's None, or `""`

Commit:
Merge remote-tracking branch 'origin/master' into adas/scim-for-roles/provision-new-role

Commit:
Applied comments, fixed errors, and applied new-new permissions check

RyanSkonnord commented on src/sentry/scim/endpoints/members.py:
```diff
@@ -372,14 +375,31 @@ def post(self, request: Request, organization) -> Response:
+            role = organization.default_role
+
+        # Allow any role as long as it doesn't have `org:admin` permissions
+        allowed_roles = {role for role in roles.get_all() if not role.has_scope("org:admin")}
```
Nice. Even though this still may be a little dicey in the long run, it's good that we're tying it to a scope and not hard-coding any assumptions about the structure of `SENTRY_ROLES`.
